### PR TITLE
Enabling hint for Optical flow calculation

### PR DESCRIPTION
**Pull request by szalpal**

### CODE REVIEW

File: dali/aux/optical_flow/turing_of/optical_flow_turing.cc [hunk 1]
```diff
@@ -66,13 +66,21 @@ OpticalFlowTuring::OpticalFlowTuring(dali::optical_flow::OpticalFlowParams param
...
+  }
 }
 
 
 OpticalFlowTuring::~OpticalFlowTuring() {
   inbuf_.reset(nullptr);
   refbuf_.reset(nullptr);
   outbuf_.reset(nullptr);
+  if (of_params_.enable_external_hints) {
+    hintsbuf_.reset(nullptr);
```
Comment (jantonguirao) [inline dali/aux/optical_flow/turing_of/optical_flow_turing.cc]:
```suggestion
    hintsbuf_.reset();
```

File: dali/pipeline/operators/util/optical_flow.h [hunk 2]
```diff
@@ -37,10 +37,11 @@ struct backend_to_compute<GPUBackend> {
   using type = kernels::ComputeGPU;
 };
 
-const std::string kPresetArgName = "preset";   // NOLINT
-const std::string kOutputFormatArgName = "output_format";   // NOLINT
-const std::string kEnableHintsArgName = "enable_hints";   // NOLINT
-const std::string kImageTypeArgName = "image_type";   // NOLINT
+const std::string kPresetArgName = "preset";                               // NOLINT
```
Comment (jantonguirao) [inline dali/pipeline/operators/util/optical_flow.h]: why NOLINT?

File: dali/aux/optical_flow/turing_of/optical_flow_turing.h [hunk 3]
```diff
@@ -82,11 +82,29 @@ DLL_PUBLIC void
 DecodeFlowComponents(const int16_t *input, float *output, size_t pitch, size_t width_px,
                      size_t height, cudaStream_t stream = 0);
 
+/**
+ * Encode flow components and put in strided memory (for external hitns)
```
Comment (Kh4L) [inline dali/aux/optical_flow/turing_of/optical_flow_turing.h]:
```suggestion
 * Encode flow components and put in strided memory (for external hints)
```

File: dali/aux/optical_flow/turing_of/optical_flow_turing.h [hunk 4]
```diff
@@ -82,11 +82,29 @@ DLL_PUBLIC void
...
+                     size_t height, cudaStream_t stream = 0);
+
 
 inline __host__ __device__ float decode_flow_component(int16_t value) {
   return value * (1 / 32.f);
 }
 
+
+inline __host__ __device__ int16_t encode_flow_component(float value) {
+  return static_cast<int16_t>(value * 32);
```
Comment (Kh4L) [inline dali/aux/optical_flow/turing_of/optical_flow_turing.h]:
```suggestion
  return static_cast<int16_t>(value * 32.f);
```
for consistency

File: dali/pipeline/operators/util/optical_flow.cc [hunk 5]
```diff
@@ -33,8 +33,11 @@ Dali uses Turing optical flow hardware implementation: https://developer.nvidia.
                                 R"code(Setting grid size for output vector.
 Value defines width of grid square (e.g. if value == 4, 4x4 grid is used).
 For values <=0, grid size is undefined. Currently only grid_size=4 is supported.)code", -1, false)
-                .AddOptionalArg(detail::kEnableHintsArgName,
+                .AddOptionalArg(detail::kEnableTemporalHintsArgName,
```
Comment (Kh4L) [inline dali/pipeline/operators/util/optical_flow.cc]: I think briefly explaining what are temporal and external hints here would be helpful

**Comment by Kh4L** [comment]
Also, you should change optical_flow.cc to optical_flow.cu, since it's a GPU implem

File: dali/pipeline/operators/util/optical_flow.h (hunk 2 above)
Comment (szalpal) [reply dali/pipeline/operators/util/optical_flow.h]: Cause linter doesn't like global `const std::string`s :(

File: dali/aux/optical_flow/turing_of/optical_flow_turing.cc (hunk 1 above)
Comment (szalpal) [reply dali/aux/optical_flow/turing_of/optical_flow_turing.cc]: Here I followed the example in `cppreference`:
```
    std::cout << "Release and delete the owned Foo...\n";
    up.reset(nullptr); 
```
I would prefer to keep it this way

File: dali/aux/optical_flow/turing_of/optical_flow_turing.h (hunk 4 above)
Comment (szalpal) [reply dali/aux/optical_flow/turing_of/optical_flow_turing.h]: Done

File: dali/pipeline/operators/util/optical_flow.cc (hunk 5 above)
Comment (szalpal) [reply dali/pipeline/operators/util/optical_flow.cc]: Done

File: dali/aux/optical_flow/turing_of/optical_flow_turing.h (hunk 3 above)
Comment (szalpal) [reply dali/aux/optical_flow/turing_of/optical_flow_turing.h]: Done

**Comment by szalpal** [comment]
#688928

**Comment by szalpal** [comment]
DALI-340

**Pull request merged**